pkg/controller/sparkapplication/controller.go

```diff
@@ -159,27 +248,84 @@ func newSparkApplicationController(
 	return controller
 }
 
+// StartQueueCleanupRoutine starts a go routine to clean up queues for SparkApplications that have not been updated for a certain period.
+func (c *Controller) StartQueueCleanupRoutine(checkInterval time.Duration, maxAge time.Duration, stopCh <-chan struct{}) {
+	ticker := time.NewTicker(checkInterval)
+	defer ticker.Stop()
+
+	// Infinite loop that listens for ticker.C channel and stopCh channel.
+	for {
+		select {
+		case <-ticker.C: // every checkInterval duration
+			c.cleanupQueues(maxAge)
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// cleanupQueues deletes the queues for SparkApplications that have not been updated for a certain period.
+func (c *Controller) cleanupQueues(maxAge time.Duration) {
+	queuesBefore := 0
+	queuesAfter := 0
+	for i := range c.appQueuesMaps {
+		currM := c.appQueuesMaps[i].appQueueMap
+		currL := c.appQueuesMaps[i].appQueueMutex
+		if len(currM) == 0 {
+			glog.V(0).Info("No need to clean up Map %d as it is empty, skipping cleanup", i)
+			continue
+		}
+		queuesBefore += len(currM)
+		now := time.Now()
+		glog.V(0).Infof("Cleaning up queues for SparkApplications at time: %v", now)
+		// we prevent deadlock we mark the quesues for deletion and delete them outside the lock
+		var queuesToDelete []string
+		currL.RLock()
+		for appName, queueInfo := range currM {
+			glog.V(0).Infof("Checking queue for app %v", appName)
+			if now.Sub(queueInfo.LastUpdateTs) > maxAge {
+				glog.V(0).Infof("Sending %v to deletion due to inactivity", appName)
+				queuesToDelete = append(queuesToDelete, appName)
+			}
+		}
+		currL.RUnlock()
+		for _, appName := range queuesToDelete {
+			c.deleteImmediateQueue(appName)
+		}
+
+		queuesAfter += len(currM)
+	}
+	glog.V(0).Infof("Cleaned up %d queues, %d remaining", queuesBefore-queuesAfter, queuesAfter)
+
+}
+
+// StopQueueCleanupRoutine stops the go routine that cleans up queues.
+func (c *Controller) StopQueueCleanupRoutine(stopCh chan struct{}) {
+	close(stopCh)
+}
+
 // Start starts the Controller by registering a watcher for SparkApplication objects.
-func (c *Controller) Start(workers int, stopCh <-chan struct{}) error {
+func (c *Controller) Start(maxAge time.Duration, checkInterval time.Duration, stopCh <-chan struct{}) error {
 	// Wait for all involved caches to be synced, before processing items from the queue is started.
 	if !cache.WaitForCacheSync(stopCh, c.cacheSynced) {
 		return fmt.Errorf("timed out waiting for cache to sync")
 	}
```

Comment:
With this change, we are breaking the leader election feature, as previously the worker thread creation is blocked by startCh so that only the leader operator has active worker thread. But we right now we are creating worker thread as new sparkapp comes in
https://www.kubeflow.org/docs/components/spark-operator/user-guide/leader-election/